### PR TITLE
requester: defer execution of request.Item generation in closure

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Requester system method SendPayload() change: Implement closure to defer *optional* signature/nonce generations on every attempt, this allows for retry attempts on exchanges which have NonceEnabled = true; this can also keep our payloads in line with time receive windows as execution occurs after rate limiting.

Applicable impacted exchanges:
* Bitfinex 
* Bithumb
* Bitstamp
* Exmo
* Gemini
* Itbit
* Kraken
* Localbitcoins
* Yobit

Opened as draft as this might be a confusing implementation across GCT, which will require feedback and implementable options. 

Testing: All services will need to be tested.

NOTE: This removes deadlines that were protecting the HTTP auth request.Item generation. This was implemented to make sure that we are still within time receive windows either by friction caused by rate limiting or by the retry/back off mechanism. In this PR we are regenerating our authenticated fields per request attempt therefore this usage of deadline context can be deprecated. A new context PR with correct context propagation from gRPC/client side requests should provide a better (timeout/deadline) policy in future [trello card](https://trello.com/c/Y92GMCgZ/125-context-propagation-from-grpc-client-with-configurable-attributes-to-request-system).

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by gloriousCode — CHANGES_REQUESTED** [review]
I like the design of this and it fixes a problem for retrying auth requests. An idea perhaps we'll want is the ability to add a no-retry on some of them, like submitting a market order when rate limited, then it does it 5 minutes later at a way worse price :D But I'm not asking for that in this PR.

File: exchanges/binance/binance.go [hunk 1]
```diff
@@ -578,9 +578,6 @@ func (b *Binance) OpenOrders(pair currency.Pair) ([]QueryOrderData, error) {
 			return nil, err
 		}
 		params.Add("symbol", p)
-	} else {
-		// extend the receive window when all currencies to prevent "recvwindow" error
-		params.Set("recvWindow", "10000")
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/binance/binance.go:583]: Yeah I will revert those changes.

File: exchanges/bittrex/bittrex.go [hunk 2]
```diff
@@ -356,7 +356,7 @@ func (b *Bittrex) SendHTTPRequest(ep exchange.URL, path string, result interface
 		HTTPRecording:  b.HTTPRecording,
 		HeaderResponse: resultHeader,
 	}
-	return b.SendPayload(context.Background(), &requestItem)
+	return b.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) { return &requestItem, nil })
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/bittrex/bittrex.go:359]: Yeah I will make them consistent; as for this we can just assign a ptr and return that so on every retry it doesn't have to do any more 'work' in that frame without needing to. e.g: 
```go
// for no required changes per request
item := &request.Item{
// fill in details
}
return b.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) { return item, nil })

// for required changes 

b.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) { 
// header signature gen off timestamp etc
return &request.Item{
// fill in details
}, nil })
```

Does that sound okay?

[the remark this answers is not in the document]
Comment (gloriousCode) [reply exchanges/bittrex/bittrex.go:359]: Sounds lovely

**Comment by shazbert** [comment]
> An idea perhaps we'll want is the ability to add a no-retry on some of them, like submitting a market order when rate limited, then it does it 5 minutes later at a way worse price :D But I'm not asking for that in this PR

That's not ideal 😆, I might be able to upgrade the rate limiter to be context aware calling (*limiter).WaitN(ctx,n) to check for any deadlines. Will investigate more.

File: exchanges/nonce/nonce.go [hunk 3]
```diff
@@ -27,8 +27,10 @@ func (n *Nonce) Get() Value {
 
 // GetInc increments and returns the value of the nonce
 func (n *Nonce) GetInc() Value {
-	n.Inc()
-	return n.Get()
+	n.m.Lock()
+	defer n.m.Unlock()
+	n.n++
+	return Value(n.n)
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/nonce/nonce.go:26]: I removed it, it doesn't seem to be used and I didn't want to double lock and unlock.

File: exchanges/request/request.go [hunk 4]
```diff
@@ -39,30 +48,22 @@ func New(name string, httpRequester *http.Client, opts ...RequesterOption) *Requ
...
-		return err
-	}
-
-	if i.HTTPDebugging {
-		// Err not evaluated due to validation check above
-		dump, _ := httputil.DumpRequestOut(req, true)
-		log.Debugf(log.RequestSys, "DumpRequest:\n%s", dump)
+	if newRequest == nil {
+		return errRequestFunctionIsNil
 	}
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/request/request.go:60]: I defered it

File: exchanges/bitmex/bitmex.go [hunk 5]
```diff
@@ -887,17 +876,20 @@ func (b *Bitmex) SendAuthenticatedHTTPRequest(ep exchange.URL, verb, path string
 
 	ctx, cancel := context.WithDeadline(context.Background(), expires)
 	defer cancel()
-	err = b.SendPayload(ctx, &request.Item{
+	fullPath := endpoint + path
```
Comment (gloriousCode) [inline exchanges/bitmex/bitmex.go:879]: This function also uses timestamp based signature and headers. I would advise inlining. Bittrex has rate limits and can be prone to expiry

File: exchanges/bittrex/bittrex.go [hunk 6]
```diff
@@ -408,6 +409,10 @@ func (b *Bittrex) SendAuthHTTPRequest(ep exchange.URL, method, action string, pa
 		HTTPDebugging:  b.HTTPDebugging,
 		HTTPRecording:  b.HTTPRecording,
 		HeaderResponse: resultHeader,
+	}
+
+	return b.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) {
```
Comment (gloriousCode) [inline exchanges/bittrex/bittrex.go:414]: This function also uses timestamp based signature and headers. I would advise inlining. Bittrex has an unimplemented rate limit (not in scope, but relevant)

File: exchanges/btse/btse.go [hunk 7]
```diff
@@ -518,7 +520,10 @@ func (b *BTSE) SendAuthenticatedHTTPRequest(ep exchange.URL, method, endpoint st
 		Verbose:       b.Verbose,
 		HTTPDebugging: b.HTTPDebugging,
 		HTTPRecording: b.HTTPRecording,
-		Endpoint:      f,
+	}
+
+	return b.SendPayload(context.Background(), f, func() (*request.Item, error) {
+		return item, nil
```
Comment (gloriousCode) [inline exchanges/btse/btse.go:526]: This function also uses timestamp based signature and headers. I would advise inlining. BTSE already has rate limits

File: exchanges/coinbasepro/coinbasepro.go [hunk 8]
```diff
@@ -742,6 +748,9 @@ func (c *CoinbasePro) SendAuthenticatedHTTPRequest(ep exchange.URL, method, path
 		Verbose:       c.Verbose,
 		HTTPDebugging: c.HTTPDebugging,
 		HTTPRecording: c.HTTPRecording,
+	}
+	return c.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
+		return item, nil
```
Comment (gloriousCode) [inline exchanges/coinbasepro/coinbasepro.go:753]: This function also uses timestamp based signature/headers. I would advise inlining. Coinbasepro has rate limits

File: exchanges/coinbene/coinbene.go [hunk 9]
```diff
@@ -1189,7 +1191,10 @@ func (c *Coinbene) SendAuthHTTPRequest(ep exchange.URL, method, path, epPath str
 		Verbose:       c.Verbose,
 		HTTPDebugging: c.HTTPDebugging,
 		HTTPRecording: c.HTTPRecording,
-		Endpoint:      f,
+	}
+
+	if err := c.SendPayload(ctx, f, func() (*request.Item, error) {
```
Comment (gloriousCode) [inline exchanges/coinbene/coinbene.go:1196]: This function also uses timestamp based signature/headers. I would advise inlining. Coinbene has rate limits

File: exchanges/huobi/huobi.go [hunk 10]
```diff
@@ -877,6 +884,10 @@ func (h *HUOBI) SendAuthenticatedHTTPRequest(ep exchange.URL, method, endpoint s
 		Verbose:       h.Verbose,
 		HTTPDebugging: h.HTTPDebugging,
 		HTTPRecording: h.HTTPRecording,
+	}
+
+	err = h.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
```
Comment (gloriousCode) [inline exchanges/huobi/huobi.go:889]: This function also uses timestamp based signature/headers. I would advise inlining. Huobi has rate limits

File: exchanges/huobi/huobi_futures.go [hunk 11]
```diff
@@ -1159,6 +1160,10 @@ func (h *HUOBI) FuturesAuthenticatedHTTPRequest(ep exchange.URL, method, endpoin
 		Verbose:       h.Verbose,
 		HTTPDebugging: h.HTTPDebugging,
 		HTTPRecording: h.HTTPRecording,
+	}
+
+	if err := h.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
```
Comment (gloriousCode) [inline exchanges/huobi/huobi_futures.go:1165]: This function also uses timestamp based signature/headers. I would advise inlining. Huobi has rate limits

File: exchanges/kraken/kraken_futures.go [hunk 12]
```diff
@@ -292,6 +293,10 @@ func (k *Kraken) SendFuturesAuthRequest(method, path string, postData url.Values
 		Verbose:       k.Verbose,
 		HTTPDebugging: k.HTTPDebugging,
 		HTTPRecording: k.HTTPRecording,
+	}
+
+	err := k.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
+		return item, nil
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_futures.go:299]: This function also uses timestamp based signature/headers. I would advise inlining. Kraken has unimplemented rate limits

File: exchanges/okgroup/okgroup.go [hunk 13]
```diff
@@ -632,11 +630,17 @@ func (o *OKGroup) SendHTTPRequest(ep exchange.URL, httpMethod, requestType, requ
 		Verbose:       o.Verbose,
 		HTTPDebugging: o.HTTPDebugging,
 		HTTPRecording: o.HTTPRecording,
+	}
+
+	err = o.SendPayload(ctx, request.Unset, func() (*request.Item, error) {
+		return item, nil
```
Comment (gloriousCode) [inline exchanges/okgroup/okgroup.go:636]: This function also uses timestamp based signature/headers. I would advise inlining. OKGroup has unimplemented rate limits (not in scope, but relevant)

File: exchanges/poloniex/poloniex.go [hunk 14]
```diff
@@ -877,6 +882,10 @@ func (p *Poloniex) SendAuthenticatedHTTPRequest(ep exchange.URL, method, endpoin
 		Verbose:       p.Verbose,
 		HTTPDebugging: p.HTTPDebugging,
 		HTTPRecording: p.HTTPRecording,
+	}
+
+	return p.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) {
+		return item, nil
```
Comment (gloriousCode) [inline exchanges/poloniex/poloniex.go:888]: This function also uses timestamp based signature/headers. I would advise inlining. Poloniex has rate limits

File: exchanges/zb/zb.go [hunk 15]
```diff
@@ -331,17 +334,18 @@ func (z *ZB) SendAuthenticatedHTTPRequest(ep exchange.URL, httpMethod string, pa
...
 		Verbose:       z.Verbose,
 		HTTPDebugging: z.HTTPDebugging,
 		HTTPRecording: z.HTTPRecording,
-		Endpoint:      f,
-	})
+	}
+
+	err = z.SendPayload(ctx, f, func() (*request.Item, error) { return item, nil })
 	if err != nil {
 		return err
```
Comment (gloriousCode) [inline exchanges/zb/zb.go:350]: This function also uses timestamp based signature/headers. I would advise inlining. ZB has rate limits

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! I'm a little confused as you'll see with my comments. Shouldn't all requests which have nonce/timestamp based signatures/headers be using the inline generation? I've noted all exchanges which don't allow for the regeneration but have either ratelimits or unimplemented rate limits (which is out of scope). So all the ones I've noted have the potential to be retried/be expired by the time the rate limit has finished or retried and so I think it would be good to put them inline.

**Comment by shazbert** [comment]
> Shouldn't all requests which have nonce/timestamp based signatures/headers be using the inline generation? 

&& 

> So all the ones I've noted have the potential to be retried/be expired by the time the rate limit has finished or retried and so I think it would be good to put them inline.



I think you might be correct here, but I didn't want to change a function state between request attempts if I didn't need to. On more thinking I will probably favour a full change from what you have requested. 

> I've noted all exchanges which don't allow for the regeneration but have either ratelimits or unimplemented rate limits (which is out of scope). 

Thanks for doing that, I can open a PR addressing these.  🦅 🚀

File: exchanges/exmo/exmo.go [hunk 16]
```diff
@@ -326,39 +330,41 @@ func (e *EXMO) SendAuthenticatedHTTPRequest(epath exchange.URL, method, endpoint
...
-	headers["Sign"] = crypto.HexEncodeToString(hash)
-	headers["Content-Type"] = "application/x-www-form-urlencoded"
-
-	path := fmt.Sprintf("/v%s/%s", exmoAPIVersion, endpoint)
+		if e.Verbose {
+			log.Debugf(log.ExchangeSys, "Sending %s request to %s with params %s\n",
+				method,
+				endpoint,
+				payload)
+		}
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/exmo/exmo.go:349]: opted to remove as we have [requester debug info on verbose](https://github.com/shazbert/gocryptotrader/blob/defer_request_gen/exchanges/request/request.go#L133-L142)

File: exchanges/binance/binance.go [hunk 17]
```diff
@@ -734,41 +743,39 @@ func (b *Binance) SendAuthHTTPRequest(ePath exchange.URL, method, path string, p
...
+
 	interim := json.RawMessage{}
+	err = b.SendPayload(ctx, f, func() (*request.Item, error) {
+		params.Set("timestamp", strconv.FormatInt(time.Now().Unix()*1000, 10))
+		signature := params.Encode()
+		hmacSigned := crypto.GetHMAC(crypto.HashSHA256, []byte(signature), []byte(b.API.Credentials.Secret))
+		hmacSignedStr := crypto.HexEncodeToString(hmacSigned)
+		headers := make(map[string]string)
+		headers["X-MBX-APIKEY"] = b.API.Credentials.Key
+		path = common.EncodeURLValues(path, params)
```
Comment (gloriousCode) [inline exchanges/binance/binance.go:758]: So I've been testing how this works. I'm forcing things to do a second attempt to see what happens. The `fullpath` variable you were using outside the inline function is required upon testing it. Or at least, resetting the `path` var per attempt.
Basically, the context and values remain when retrying, so a `path = common.EncodeURLValues(path, params)` will re-append the values leading to the following type of output where values are duplicated:
```
https://api.binance.com/wapi/v3/depositAddress.html?asset=BTC&recvWindow=10000&status=true&timestamp=1628039926000&signature=569b7f637693510b548d2f5e1a36a5b2e6485de914eee188b5d3f751c920d5e2?asset=BTC&recvWindow=10000&status=true&timestamp=1628039949000&signature=3f9b91345982626a3eaf6d588d8ea6889383051a477599b072b33027bd54c399
```
I'm seeing this style appending on Binance, Bitfinex and Coinbene

I've tested forcing an auth retry on FTX and it works well 🎉

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thank you for making those changes! There are some lint issues and I came across an error during testing, but otherwise its looking good! 🎉

File: exchanges/binance/binance.go (hunk 17 above)
Comment (shazbert) [reply exchanges/binance/binance.go:758]: 🦅 Will fix! Thanks.

File: exchanges/bitstamp/bitstamp.go [hunk 18]
```diff
@@ -617,59 +620,56 @@ func (b *Bitstamp) SendAuthenticatedHTTPRequest(ep exchange.URL, path string, v2
...
+		values.Set("nonce", n)
+		hmac := crypto.GetHMAC(crypto.HashSHA256,
+			[]byte(n+b.API.Credentials.ClientID+b.API.Credentials.Key),
+			[]byte(b.API.Credentials.Secret))
+		values.Set("signature", strings.ToUpper(crypto.HexEncodeToString(hmac)))
+
+		if v2 {
+			path = endpoint + "/v" + bitstampAPIVersion + "/" + path + "/"
+		} else {
+			path = endpoint + "/" + path + "/"
```
Comment (gloriousCode) [inline exchanges/bitstamp/bitstamp.go:642]: Path will need to be adapted here too

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes! Found two more instances going over each

File: exchanges/kraken/kraken_futures.go [hunk 19]
```diff
@@ -264,40 +264,45 @@ func (k *Kraken) SendFuturesAuthRequest(method, path string, postData url.Values
...
+	interim := json.RawMessage{}
+	newRequest := func() (*request.Item, error) {
+		nonce := strconv.FormatInt(time.Now().UnixNano()/1000000, 10)
+		reqData := ""
+		if len(data) > 0 {
+			temp, err := json.Marshal(data)
+			if err != nil {
+				return nil, err
+			}
+			postData.Add("json", string(temp))
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/kraken/kraken_futures.go:277]: 🦅 !

File: exchanges/localbitcoins/localbitcoins.go [hunk 20]
```diff
@@ -751,43 +754,36 @@ func (l *LocalBitcoins) SendAuthenticatedHTTPRequest(ep exchange.URL, method, pa
...
+		message := n + l.API.Credentials.Key + path + encoded
+		hmac := crypto.GetHMAC(crypto.HashSHA256, []byte(message), []byte(l.API.Credentials.Secret))
+		headers := make(map[string]string)
+		headers["Apiauth-Key"] = l.API.Credentials.Key
+		headers["Apiauth-Nonce"] = n
+		headers["Apiauth-Signature"] = strings.ToUpper(crypto.HexEncodeToString(hmac))
+		headers["Content-Type"] = "application/x-www-form-urlencoded"
+
+		if method == http.MethodGet && len(encoded) > 0 {
+			path += "?" + encoded
```
Comment (gloriousCode) [inline exchanges/localbitcoins/localbitcoins.go:772]: While I don't have API keys, `path` appending path will cause an issue on retries

File: exchanges/zb/zb.go [hunk 21]
```diff
@@ -312,40 +315,38 @@ func (z *ZB) SendAuthenticatedHTTPRequest(ep exchange.URL, httpMethod string, pa
...
+		params.Set("sign", fmt.Sprintf("%x", hmac))
+
+		urlPath := fmt.Sprintf("%s/%s?%s",
+			endpoint,
+			params.Get("method"),
+			params.Encode())
+
+		return &request.Item{
+			Method:        httpMethod,
+			Path:          urlPath,
```
Comment (gloriousCode) [inline exchanges/zb/zb.go:331]: The API url now appears to be `http://api.zb.com` now, not `http://api.zb.live`. This is undocumented however. But I do get successful responses from there

**Review by gloriousCode — CHANGES_REQUESTED** [review]
I experienced the following issues testing every exchange.
Note, I could not test the following:
- Bithumb
  - they do not send out API key activation email
- Bitstamp
  - They do not send out API key activation email
- Itbit
  - Cannot access new owner's service
- Localbitcoin
  - No api key
- Okgroup
  - v3 api key expired and can only generate v5 keys. Hopefully you have tested it.

File: exchanges/gateio/gateio.go [hunk 22]
```diff
@@ -431,6 +434,9 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ep exchange.URL, method, endpoint,
 		Verbose:       g.Verbose,
 		HTTPDebugging: g.HTTPDebugging,
 		HTTPRecording: g.HTTPRecording,
+	}
+	err = g.SendPayload(context.Background(), request.Unset, func() (*request.Item, error) {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/gateio.go:438]: Unfortunately you cannot read from the same state twice. Nice spotting 🦅 ! Fixed now.

**Comment by shazbert** [comment]
@gloriousCode I went over the exchanges you could not do to verify and make sure:

Bithumb - Tested and works
Bitstamp - Tested and works
Itbit - Cannot generate new key, cannot verify but I can verify that no appends occur on path.
Localbitcoin - Tested and works 
Okgroup - Like you said v3 api key expired and can only generate v5 keys. We need to upgrade to v5 endpoints because they have purged all my prior keys on update.

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes! Can now redo requests 🎉

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for handling those extra issues

**Review by thrasher- — APPROVED** [review]
Tested auth requests based on the the following exchanges:

Bitfinex
Bitstamp
Kraken

Also ran the wrapper tool to ensure that changes to the requester didn't break other exchanges functionality

Looking forward to the addition of https://github.com/thrasher-corp/gocryptotrader/pull/731 !

**Pull request merged**